insonmnia/marketplace/server.go

```diff
@@ -0,0 +1,263 @@
+package marketplace
+
+import (
+	"errors"
+	"sync"
+
+	"net"
+
+	"github.com/pborman/uuid"
+	pb "github.com/sonm-io/core/proto"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+var (
+	errOrderNotFound  = errors.New("Order cannot be found")
+	errPriceIsZero    = errors.New("Order price cannot be less or equal than zero")
+	errOrderIsNil     = errors.New("Order cannot be nil")
+	errSlotIsNil      = errors.New("Order slot cannot be nil")
+	errResourcesIsNil = errors.New("Slot resources cannot be nil")
+)
+
+type OrderStorage interface {
+	GetOrders(slot *pb.Slot) ([]*pb.Order, error)
+	GetOrderByID(id string) (*pb.Order, error)
+	CreateOrder(order *pb.Order) (*pb.Order, error)
+	DeleteOrder(id string) error
+}
+
+type inMemOrderStorage struct {
+	sync.RWMutex
+	db map[string]*pb.Order
+}
+
+func (in *inMemOrderStorage) generateID() string {
+	return uuid.New()
+}
+
+func (in *inMemOrderStorage) validateOrder(order *pb.Order) error {
+	// todo: check that order struct has all required fields
+	if order == nil {
+		return errOrderIsNil
+	}
+
+	err := in.validateSlot(order.Slot)
+	if err != nil {
+		return err
+	}
+
+	if order.Price <= 0 {
+		return errPriceIsZero
+	}
+
+	return nil
+}
+
+func (in *inMemOrderStorage) validateSlot(slot *pb.Slot) error {
+	if slot == nil {
+		return errSlotIsNil
+	}
+
+	if slot.Resources == nil {
+		return errResourcesIsNil
+	}
+	return nil
+}
+
+func (in *inMemOrderStorage) compareTime(slot *pb.Slot, order *pb.Order) bool {
```

Comment:
Order contains a slot, which means that you're comparing a slot with another slot. Then maybe better to represent it explicitly?

`func (s *Slot) compareTime(other *Slot) bool` and so on

Reply:
Of course you need a wrapper type: `type Slot pb.Slot`.